modules/masterbar/admin-menu/class-admin-menu.php

```diff
@@ -0,0 +1,736 @@
+	}
+
+	/**
+	 * Adds Media menu.
+	 *
+	 * @param string $domain  Site domain.
+	 * @param bool   $calypso Optional. Whether links should point to Calypso or wp-admin. Default true (Calypso).
+	 */
+	public function add_media_menu( $domain, $calypso = true ) {
+		remove_submenu_page( 'upload.php', 'upload.php' );
+		remove_submenu_page( 'upload.php', 'media-new.php' );
+
+		if ( $calypso ) {
+			$menu_slug = 'https://wordpress.com/media/' . $domain;
+
+			remove_menu_page( 'upload.php' );
+			add_menu_page( __( 'Media', 'jetpack' ), __( 'Media', 'jetpack' ), 'upload_files', $menu_slug, null, 'dashicons-admin-media', 10 );
+			$this->migrate_submenus( 'upload.php', $menu_slug );
+		}
+	}
+
+	/**
+	 * Adds Page menu.
+	 *
+	 * @param string $domain  Site domain.
+	 * @param bool   $calypso Optional. Whether links should point to Calypso or wp-admin. Default true (Calypso).
+	 */
+	public function add_page_menu( $domain, $calypso = true ) {
+		if ( ! $calypso ) {
+			return;
+		}
+
+		$ptype_obj = get_post_type_object( 'page' );
+		$menu_slug = 'https://wordpress.com/pages/' . $domain;
+
+		remove_menu_page( 'edit.php?post_type=page' );
+		remove_submenu_page( 'edit.php?post_type=page', 'edit.php?post_type=page' );
+		remove_submenu_page( 'edit.php?post_type=page', 'post-new.php?post_type=page' );
+
+		add_menu_page( esc_attr( $ptype_obj->labels->menu_name ), $ptype_obj->labels->menu_name, $ptype_obj->cap->edit_posts, $menu_slug, null, 'dashicons-admin-page', $ptype_obj->menu_position );
+		add_submenu_page( $menu_slug, $ptype_obj->labels->all_items, $ptype_obj->labels->all_items, $ptype_obj->cap->edit_posts, $menu_slug, null, 5 );
+		add_submenu_page( $menu_slug, $ptype_obj->labels->add_new, $ptype_obj->labels->add_new, $ptype_obj->cap->create_posts, 'https://wordpress.com/page/' . $domain, null, 10 );
+		$this->migrate_submenus( 'edit.php?post_type=page', $menu_slug );
+	}
+
+	/**
+	 * Adds custom post type menus.
+	 *
+	 * @param \WP_Post_Type $ptype_obj Post type object.
+	 * @param string        $domain    Site domain.
+	 * @param bool          $calypso   Optional. Whether links should point to Calypso or wp-admin.
+	 *                                 Default true (Calypso).
+	 */
+	public function add_custom_post_type_menu( \WP_Post_Type $ptype_obj, $domain, $calypso = true ) {
+		if ( ! $calypso ) {
+			return;
+		}
+
+		if ( ! $ptype_obj->show_in_menu ) {
+			return;
+		}
+
+		$post_type = $ptype_obj->name;
+		$menu_icon = 'dashicons-admin-post';
+		if ( is_string( $ptype_obj->menu_icon ) ) {
+			// Special handling for data:image/svg+xml and Dashicons.
+			if ( 0 === strpos( $ptype_obj->menu_icon, 'data:image/svg+xml;base64,' ) || 0 === strpos( $ptype_obj->menu_icon, 'dashicons-' ) ) {
+				$menu_icon = $ptype_obj->menu_icon;
+			} else {
+				$menu_icon = esc_url( $ptype_obj->menu_icon );
+			}
+		}
+
+		$old_slug   = 'edit.php?post_type=' . $post_type;
+		$ptype_file = 'https://wordpress.com/types/' . $post_type . '/' . $domain;
```

Comment:
@davemart-in Jeremy is making a good point here. I'm not sure you and I discussed the default behavior for custom post types so far and I defaulted to Calypso. Should we limit Calypso links to posts and pages for now?

Reply:
> I wonder if we shouldn't favor wp-admin for Custom Post Type management in general.
> Should we limit Calypso links to posts and pages for now?

@obenland, @jeherve What are the implications of this? I'm not sure I fully understand the pros and cons.

Reply:
The biggest challenge I see with editing custom post types (CPT) in Calypso by default is that CPTs can remove support for the editor for example, or just be generally incompatible with Gutenberg. Jeremy also encountered problems with capabilities it seems.

The more I think about it, the more sense it makes to me to just keep those in wp-admin for now and limit calypso links to posts and pages.

Reply:
Konstantin just clarified via DM with:
> For menu items of custom post types, keeping them in wp-admin rather than linking them to Calypso. So have them point to example.org/wp-admin/edit.php?post_type=custom_post_type, rather than wp.com/types/custom_post_type/example.org

That is the piece I didn't understand before. Yes, I agree that we should do this for CPT's.

Reply:
9f6cc03